vignettes/civis_scripts.Rmd

```diff
@@ -0,0 +1,296 @@
+
+## Creating and Running Scripts
+
+Let's make these concepts concrete with an example! We'll use the 'R' language script throughout,
+but `container` scripts work exactly the same way. In the second section,
+we'll cover `custom` and `template` scripts.
+
+### An Example Script
+
+The `post` method creates the job and returns a list of metadata about it, including it's type.
+
+```{r, eval = FALSE}
+source <- c('
+ print("Hello World!")
+')
+job <- scripts_post_r(name = 'Hello!', source = source)
+```
+
+Each script can be uniquely identified by its _job id_. If you have a job id 
+but don't know what kind of script it is, you can do `jobs_get(id)`. 
+
+Each script type is associated with its own API endpoints. For instance, to post a job of each script type, you need `scripts_post_r`, `scripts_post_containers`, `scripts_post_custom`, or `templates_post_scripts`. 
+
+This job hasn't been run yet. To kick off a run do:
+
+```{r, eval = FALSE}
+run <- scripts_post_r_runs(job$id)
+
+# check the status
+scripts_get_r_runs(job$id, run$id)
+
+# automatically poll until the job completes
+await(scripts_get_r_runs, id = job$id, run_id = run$id)
+```
+
+Since kicking off a job and polling until it completes is 
+a really common task for this guide, let's make it a function:
+
+```{r}
+run_script <- function(source, name = 'Cool') {
+  job <- scripts_post_r(name = name, source = source)
+  run <- scripts_post_r_runs(job$id)
+  await(scripts_get_r_runs, id = job$id, run_id = run$id)
+}
+```
+
+### Run Outputs
+
+This script isn't very useful because it doesn't produce any output that we can access. 
+To add an output to a job, we can use `scripts_post_r_runs_outputs`. Notice
+that the environment variables `CIVIS_JOB_ID` and `CIVIS_RUN_ID` are 
+automatically inserted into the environment for us to have access to.
+
+```{r, eval=FALSE}
+source <- c("
+ library(civis)
+ data(iris)
+ write.csv(iris, 'iris.csv')
+ job_id <- as.numeric(Sys.getenv('CIVIS_JOB_ID'))
+ run_id <- as.numeric(Sys.getenv('CIVIS_RUN_ID'))
+ file_id <- write_civis_file('iris.csv')
+ scripts_post_r_runs_outputs(job_id, run_id, object_type = 'File', object_id = file_id)
+")
+run <- run_script(source)
+```
+
+Since this pattern is so common, we replaced it with the function `write_job_output` which 
+you can use to post a filename as a run output for any script type.
+
+```{r, eval=FALSE}
+source <- c("
+ library(civis)
+ data(iris)
+ write.csv(iris, 'iris.csv')
+ write_job_output('iris.csv')
+")
+run <- run_script(source)
+```
+
+To retrieve script outputs we can use `scripts_list_r_runs_outputs`:
+
+```{r}
+out <- scripts_list_r_runs_outputs(run$rId, run$id)
+iris <- read_civis(out$objectId, using = read.csv)
+```
+
+Since this pattern is also common, you can simply use `read_civis` directly.
+This will work for any script type:
+
+```{r, eval = FALSE}
+iris <- read_civis(civis_script(run$rId), using = read.csv)
```

Comment:
is `run$rID` a typo?

Reply:
Haha, it's super weird, but it's not a typo. The output of a `scripts_get_r_runs` is:

```
<civis_api>
List of 7
 $ id               : int 123685316
 $ rId              : int 22234332
 $ state            : chr "succeeded"
 $ isCancelRequested: logi FALSE
 $ startedAt        : chr "2019-03-05T15:30:53.000Z"
 $ finishedAt       : chr "2019-03-05T15:31:15.000Z"
 $ error            : NULL
 - attr(*, "class")= chr [1:2] "civis_api" "list"
```
Where `id` is the run, and `rId` is the job, of an `r` script.